src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -27,178 +25,78 @@ public class CompilerCache : ICompilerCache
         /// Initializes a new instance of <see cref="CompilerCache"/> populated with precompiled views
         /// discovered using <paramref name="provider"/>.
         /// </summary>
-        /// <param name="assemblyProvider">The <see cref="IAssemblyProvider"/> that provides assemblies
-        /// for precompiled view discovery.</param>
+        /// <param name="razorFileInfoCollections">The sequence of <see cref="RazorFileInfoCollection"/> that provides
+        /// information for precompiled view discovery.</param>
         /// <param name="loaderContextAccessor">The <see cref="IAssemblyLoadContextAccessor"/>.</param>
         /// <param name="optionsAccessor">An accessor to the <see cref="RazorViewEngineOptions"/>.</param>
-        public CompilerCache(IAssemblyProvider assemblyProvider,
-                             IAssemblyLoadContextAccessor loadContextAccessor,
-                             IOptions<RazorViewEngineOptions> optionsAccessor)
-            : this(GetFileInfos(assemblyProvider.CandidateAssemblies),
-                  loadContextAccessor.GetLoadContext(RazorFileInfoCollectionType.Assembly),
+        public CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
```

Comment:
@Alxandr, for the time we have a hybrid of what @loudej mentioned in the design notes in #2551. This approach lets us get away without having to introduce multiple `IRazorPageFactory`s or moving it into Options to allow for reordering instance. With the exception of scenario 3 from your design proposal, we do hit all the other bullet points.